### PR TITLE
adding route support

### LINKED ISSUE #392: Implement routes using abstraction layer
Routes are reconciled by HCG and processed for custom host validation, generated host replacement and workload migration.

## custom hosts enabled
- If the custom domain is not yet verified, the route will have it's host modified in place.
- If the custom domain is verified, the route will be "shadowed" to a second ingress in order to maintain the generated host which will have an owner ref of the original route.
## custom hosts disabled
- The routes are updated with a generated host

## Advanced scheduling
Confirm migration of workloads using routes still functions, and make changes as required.

## e2e tests
covered in #401

### PR DESCRIPTION
**Pull request by philbrookes**
Closes #392 

## Description of Changes
Full implementation of routes

## Remaining Work 
- Add tests
- Add metrics
- Merge route controller and ingress controller into a single controller

## Release and Testing
* Start GLBC with CRC and e2e-mock DNS provider
* kubectl apply -f samples/echo-service/dns-mock.yaml (creates mock A record for crc router hostname)
* kubectl apply -f samples/echo-service/echo-route.yaml
* See route has generated host in spec
* See route functions using curl (ignoring bad cert on local dev)
* kubectl apply -f samples/echo-service/dns-mock-approved-domain.yaml
* See domain verification goes to verified
* See route has custom host in spec
* See shadow route created using generated host
* see both hosts function using curl
* kubectl delete -f samples/echo-service/echo-route.yaml
* See shadow and DNS resources/entries are cleaned up

### CODE REVIEW

File: pkg/reconciler/dns/dns_record.go [hunk 1]
```diff
@@ -31,8 +29,6 @@ const (
 )
 
 func (c *Controller) reconcile(ctx context.Context, dnsRecord *v1.DNSRecord) error {
-	c.Logger.V(3).Info("starting reconcile of dnsRecord ", "name", dnsRecord.Name, "namespace", dnsRecord.Namespace, "cluster", logicalcluster.From(dnsRecord))
```
Comment (maleck13) [inline pkg/reconciler/dns/dns_record.go:34]: why did we remove?

File: pkg/reconciler/domainverification/reconciler.go [hunk 2]
```diff
@@ -93,7 +91,6 @@ func (dsr *domainVerificationStatus) ensureDomainVerificationStatus(ctx context.
 }
 
 func (c *Controller) reconcile(ctx context.Context, domainVerification *v1.DomainVerification) error {
-	c.Logger.V(3).Info("starting reconcile of domainVerification ", "name", domainVerification.Name, "namespace", domainVerification.Namespace, "cluster", logicalcluster.From(domainVerification))
```
Comment (maleck13) [inline pkg/reconciler/domainverification/reconciler.go:96]: same as above?

File: pkg/reconciler/dns/dns_record.go (hunk 1 above)
Comment (philbrookes) [reply pkg/reconciler/dns/dns_record.go:34]: oh, I'll put this back, I didn't intend to commit this.

File: pkg/reconciler/domainverification/reconciler.go (hunk 2 above)
Comment (philbrookes) [reply pkg/reconciler/domainverification/reconciler.go:96]: same

File: pkg/reconciler/deployment/deployment.go [hunk 3]
```diff
@@ -6,14 +6,11 @@ import (
...
 	appsv1 "k8s.io/api/apps/v1"
 
-	"github.com/kcp-dev/logicalcluster/v2"
-
 	"github.com/kuadrant/kcp-glbc/pkg/util/metadata"
 	"github.com/kuadrant/kcp-glbc/pkg/util/workloadMigration"
 )
 
 func (c *Controller) reconcile(ctx context.Context, deployment *appsv1.Deployment) error {
-	c.Logger.V(3).Info("starting reconcile of deployment ", "name", deployment.Name, "namespace", deployment.Namespace, "cluster", logicalcluster.From(deployment))
```
Comment (mikenairn) [inline pkg/reconciler/deployment/deployment.go:16]: Any reason to remove all these logs? They shouldn't get printed unless you increase the log level anyway (--zap-log-level=3)

File: pkg/reconciler/route/route_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	trafficReconcilers "github.com/kuadrant/kcp-glbc/pkg/traffic/reconcilers"
+	"github.com/kuadrant/kcp-glbc/pkg/util/workloadMigration"
+
+	utilserrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/kuadrant/kcp-glbc/pkg/util/metadata"
+)
+
+const (
+	cascadeCleanupFinalizer = "kcp.dev/cascade-cleanup"
```
Comment (mikenairn) [inline pkg/reconciler/route/route_reconciler.go:21]: Ingress uses `kaudrant.dev/cascade-cleanup` https://github.com/kcp-dev/kcp-glbc/pull/434/files#diff-ce8d5c85955bdfb1d82cb53468ae11d92b69a7c8cb083ade95ca0cda637951fbR22, shoudl these be the same?

File: pkg/reconciler/deployment/deployment.go (hunk 3 above)
Comment (philbrookes) [reply pkg/reconciler/deployment/deployment.go:16]: I thought I had restored them all, I must have missed this one

File: pkg/reconciler/route/route_reconciler.go (hunk 4 above)
Comment (philbrookes) [reply pkg/reconciler/route/route_reconciler.go]: yep, centralized this to the traffic package now

File: pkg/reconciler/service/service.go [hunk 5]
```diff
@@ -6,14 +6,11 @@ import (
...
 	corev1 "k8s.io/api/core/v1"
 
-	"github.com/kcp-dev/logicalcluster/v2"
-
 	"github.com/kuadrant/kcp-glbc/pkg/util/metadata"
 	"github.com/kuadrant/kcp-glbc/pkg/util/workloadMigration"
 )
 
 func (c *Controller) reconcile(ctx context.Context, service *corev1.Service) error {
-	c.Logger.V(3).Info("starting reconcile of service ", "name", service.Name, "namespace", service.Namespace, "cluster", logicalcluster.From(service))
```
Comment (mikenairn) [inline pkg/reconciler/service/service.go:16]: here also

File: pkg/reconciler/secret/secret.go [hunk 6]
```diff
@@ -6,14 +6,11 @@ import (
...
 	corev1 "k8s.io/api/core/v1"
 
-	"github.com/kcp-dev/logicalcluster/v2"
-
 	"github.com/kuadrant/kcp-glbc/pkg/util/metadata"
 	"github.com/kuadrant/kcp-glbc/pkg/util/workloadMigration"
 )
 
 func (c *Controller) reconcile(ctx context.Context, secret *corev1.Secret) error {
-	c.Logger.V(3).Info("starting reconcile of secret ", "name", secret.Name, "namespace", secret.Namespace, "cluster", logicalcluster.From(secret))
```
Comment (mikenairn) [inline pkg/reconciler/secret/secret.go:16]: here also

**Review by mikenairn — APPROVED** [review]

**Pull request merged**